### PR TITLE
Canonical JSON output: always sort keys on output

### PR DESCRIPTION
**Pull request by tzz**
This is a very minor internal change that ensures that in JSON output, JSON Objects' elements are always sorted by property name.  This is known as "canonical" JSON and, while a tiny bit slower than unsorted output, is much more consistent.

It breaks several tests that were relying on the wrong order, I'll push a followup commit to fix those.

### CODE REVIEW

**Comment by tzz** [comment]
Examples and tests have been adjusted.  I hope this is acceptable, it will make acceptance tests much better.

**Comment by kacfengine** [comment]
Fully agree, this will make testing much easier, and should have virtually no effect for users (and if it does, a feature release is definitely the time to change it).

File: libutils/json.c [hunk 1]
```diff
@@ -1401,9 +1407,17 @@ void JsonObjectWriteCompact(Writer *writer, const JsonElement *object)
 
     for (size_t i = 0; i < object->container.children->length; i++)
     {
+        assert(object->container.children->data[i]->propertyName);
+    }
+
+    // sort the children Seq so the output is canonical (keys are sorted)
+    // we've already asserted that the children have a valid propertyName
+    SeqSort(object->container.children, (SeqItemComparator)JsonElementPropertyCompare, NULL);
```
Comment (kacfengine) [inline libutils/json.c]: What about ``JsonObjectWrite``? We want it there too, no?

Comment (tzz) [inline libutils/json.c]: Ah, right, thank you.  Fixed and tests updated.

**Comment by tzz** [comment]
I've actually wanted this for a while, so it's great to make it happen finally.  Thanks for the review,

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
<pre>
10:33:25   CC     json.lo
10:33:25 json.c: In function ‘JsonObjectWrite’:
10:33:25 json.c:1304: warning: dereferencing ‘void *’ pointer
10:33:25 json.c:1304: error: request for member ‘propertyName’ in something not a structure or union
10:33:25 json.c: In function ‘JsonObjectWriteCompact’:
10:33:25 json.c:1419: warning: dereferencing ‘void *’ pointer
10:33:25 json.c:1419: error: request for member ‘propertyName’ in something not a structure or union
10:33:25 make[3]: *** [json.lo] Error 1
</pre>

**Comment by tzz** [comment]
Huh, I didn't see that error on my side.

**Comment by tzz** [comment]
OK, fixed.  I also use `JsonSort()` which is a `SeqSort()` wrapper, hitherto unused so I had not seen it!  @sigurdteigen was a visionary :)

**Comment by tzz** [comment]
trigger build

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
<pre>
==================================================
Starting test: json_test.c
==================================================
test_new_delete: Starting test
test_new_delete: Test completed successfully.
test_object_duplicate_key: Starting test
test_object_duplicate_key: Test completed successfully.
test_show_string: Starting test
test_show_string: Test completed successfully.
test_show_object_simple: Starting test
test_show_object_simple: Test completed successfully.
test_show_object_escaped: Starting test
test_show_object_escaped: Test completed successfully.
test_show_object_numeric: Starting test
"{
  "real": 1234.5678,
  "int": -1234567890
}" != "{
  "int": -1234567890,
  "real": 1234.5678
}"
ERROR: json_test.c:135 Failure!
test_show_object_numeric: Test failed.
test_show_object_boolean: Starting test
test_show_object_boolean: Test completed successfully.
test_show_object_compound: Starting test
"{
  "first": "one",
  "second": {
    "third": "three"
  },
  "fourth": {
    "fifth": "five"
  }
}" != "{
  "first": "one",
  "fourth": {
    "fifth": "five"
  },
  "second": {
    "third": "three"
  }
}"
ERROR: json_test.c:183 Failure!
test_show_object_compound: Test failed.
test_show_object_compound_compact: Starting test
"{"first":"one","second":{"third":"three"},"fourth":{"fifth":"five"}}" != "{"first":"one","fourth":{"fifth":"five"},"second":{"third":"three"}}"
ERROR: json_test.c:214 Failure!
test_show_object_compound_compact: Test failed.
test_show_object_array: Starting test
test_show_object_array: Test completed successfully.
test_show_array: Starting test
test_show_array: Test completed successfully.
test_show_array_compact: Starting test
test_show_array_compact: Test completed successfully.
test_show_array_boolean: Starting test
test_show_array_boolean: Test completed successfully.
test_show_array_numeric: Starting test
test_show_array_numeric: Test completed successfully.
test_show_array_object: Starting test
test_show_array_object: Test completed successfully.
test_show_array_empty: Starting test
test_show_array_empty: Test completed successfully.
test_show_array_nan: Starting test
test_show_array_nan: Test completed successfully.
test_show_array_infinity: Starting test
test_show_array_infinity: Test completed successfully.
test_object_get_string: Starting test
test_object_get_string: Test completed successfully.
test_object_get_array: Starting test
test_object_get_array: Test completed successfully.
test_object_iterator: Starting test
test_object_iterator: Test completed successfully.
test_iterator_current: Starting test
test_iterator_current: Test completed successfully.
test_array_get_string: Starting test
test_array_get_string: Test completed successfully.
test_array_iterator: Starting test
test_array_iterator: Test completed successfully.
test_copy_compare: Starting test
test_copy_compare: Test completed successfully.
test_select: Starting test
test_select: Test completed successfully.
test_merge_array: Starting test
test_merge_array: Test completed successfully.
test_merge_object: Starting test
test_merge_object: Test completed successfully.
test_parse_empty_string: Starting test
test_parse_empty_string: Test completed successfully.
test_parse_escaped_string: Starting test
test_parse_escaped_string: Test completed successfully.
test_parse_empty_containers: Starting test
test_parse_empty_containers: Test completed successfully.
test_parse_object_simple: Starting test
test_parse_object_simple: Test completed successfully.
test_parse_array_simple: Starting test
test_parse_array_simple: Test completed successfully.
test_parse_object_compound: Starting test
test_parse_object_compound: Test completed successfully.
test_parse_object_diverse: Starting test
test_parse_object_diverse: Test completed successfully.
test_parse_array_object: Starting test
test_parse_array_object: Test completed successfully.
test_parse_good_numbers: Starting test
test_parse_good_numbers: Test completed successfully.
test_parse_bad_numbers: Starting test
test_parse_bad_numbers: Test completed successfully.
test_parse_trim: Starting test
test_parse_trim: Test completed successfully.
test_parse_array_extra_closing: Starting test
test_parse_array_extra_closing: Test completed successfully.
test_parse_array_diverse: Starting test
test_parse_array_diverse: Test completed successfully.
test_parse_bad_apple2: Starting test
test_parse_bad_apple2: Test completed successfully.
test_parse_object_garbage: Starting test
test_parse_object_garbage: Test completed successfully.
test_parse_object_nested_garbage: Starting test
test_parse_object_nested_garbage: Test completed successfully.
test_parse_array_garbage: Starting test
test_parse_array_garbage: Test completed successfully.
test_parse_array_nested_garbage: Starting test
test_parse_array_nested_garbage: Test completed successfully.
test_parse_object_escaped: Starting test
test_parse_object_escaped: Test completed successfully.
test_parse_tzz_evil_key: Starting test
test_parse_tzz_evil_key: Test completed successfully.
test_parse_primitives: Starting test
test_parse_primitives: Test completed successfully.
test_array_remove_range: Starting test
test_array_remove_range: Test completed successfully.
test_remove_key_from_object: Starting test
test_remove_key_from_object: Test completed successfully.
test_detach_key_from_object: Starting test
test_detach_key_from_object: Test completed successfully.
test_parse_array_double_and_trailing_commas: Starting test
test_parse_array_double_and_trailing_commas: Test completed successfully.
test_parse_array_comma_after_brace: Starting test
test_parse_array_comma_after_brace: Test completed successfully.
test_parse_array_bad_nested_elems: Starting test
test_parse_array_bad_nested_elems: Test completed successfully.
test_parse_object_double_colon: Starting test
test_parse_object_double_colon: Test completed successfully.
test_parse_object_double_and_trailing_comma: Starting test
test_parse_object_double_and_trailing_comma: Test completed successfully.
3 out of 57 tests failed!
    test_show_object_numeric
    test_show_object_compound
    test_show_object_compound_compact
</pre>

**Comment by tzz** [comment]
Unit test `json_test.c` fixed.

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
More failures:

* ./01_vars/02_functions/data_regextract.cf.data_regextract.cf
* ./04_examples/outputs/check_outputs.cf.check_outputs.cf
* ./08_commands/01_modules/set-context.cf.set-context.cf

You should probably run through the complete set to save some time. :-)

**Comment by tzz** [comment]
I did!  The `data_regextract` issues is another cross-branch bug, sorry.  I rebased and will get it patched up.

**Comment by tzz** [comment]
All tests pass now.  Please note that `/01_vars/04_containers/parsejson_unicode.cf` was already broken, so I've patched it but @nickanderson or someone else should take a look.

**Comment by kacfengine** [comment]
Ah ok, tricky. I guess a lot is happening in core nowadays.

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
More failures in Enterprise. Damn, we are using a lot of JSON in tests! Continuing testing on private branch.

**Comment by kacfengine** [comment]
Merged together with https://github.com/cfengine/enterprise/commit/a7ba438b2ebb6bd92c117d393522a821ce746176.

**Pull request merged**